buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/RateLimitInternalExtension.java

```diff
@@ -0,0 +1,98 @@
+package org.elasticsearch.gradle.testclusters;
+
+import org.gradle.api.Project;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This extensions was meant to be used internally by testclusters
+ *
+ * It holds synchronization primitives needed to implement the rate limiting.
+ * This is tricky because we can't use Gradle workers as there's no way to make sure that tests and their clusters are
+ * allocated atomically, so we could be in a situation where all workers are tests waiting for clusters to start up.
+ *
+ * Also auto configures cleanup of executors to make sure we don't leak threads in the daemon.
+ */
+public class RateLimitInternalExtension {
+
+    private static final int EXECUTOR_SHUTDOWN_TIMEOUT = 1;
+    private static final TimeUnit EXECUTOR_SHUTDOWN_TIMEOUT_UNIT = TimeUnit.MINUTES;
+
+    private static final Logger logger =  Logging.getLogger(RateLimitInternalExtension.class);
+
+    private final Semaphore globalSemaphore;
+    private final ExecutorService executorService;
+    private final TestClusterCleanupOnShutdown cleanupThread;
+
+    public RateLimitInternalExtension() {
+        globalSemaphore = new Semaphore(maxPermits());
+        executorService = Executors.newFixedThreadPool(
+            RateLimitInternalExtension.maxPermits() +
+                1
+        );
+        cleanupThread = new TestClusterCleanupOnShutdown();
+        executorService.submit(cleanupThread);
+    }
+
+    public static Semaphore semaphore(Project project) {
+        return getSelfAsExtension(project).globalSemaphore;
+    }
+
+    public static ExecutorService executorService(Project project) {
+        return getSelfAsExtension(project).executorService;
+    }
+
+    public static TestClusterCleanupOnShutdown cleanupThread(Project project) {
+        return getSelfAsExtension(project).cleanupThread;
+    }
+
+    static int maxPermits() {
+        return Optional.ofNullable(System.getProperty("testclusters.max-nodes"))
+            .map(Integer::valueOf)
+            .orElse(Runtime.getRuntime().availableProcessors() / 2);
+    }
+
+    private static RateLimitInternalExtension getSelfAsExtension(Project project) {
+        RateLimitInternalExtension ext = project.getRootProject().getExtensions().findByType(RateLimitInternalExtension.class);
+        if (ext == null) {
+            // Configure the extension on the root project so we have a single instance per run
+            RateLimitInternalExtension newExt = project.getRootProject().getExtensions().create(
+                "__testclusters_rate_limit"  + UUID.randomUUID().toString(),
+                RateLimitInternalExtension.class
+            );
+            Thread shutdownHook = new Thread(newExt.cleanupThread::run);
```

Comment:
We have similar type of cleanup logic in a couple of places now. I feel like we should consolidate this.

Reply:
We used to have it in the plugin, but it's all moved here now. I was under the impression that I have consolidate all uses, could you help point out what I missed ?

Reply:
I think you are right. Looks like this logic was moved, not added.